ballista/scheduler/src/state/execution_graph_dot.rs

```diff
@@ -548,87 +549,75 @@ filter_expr="]
-		stage_4_0_0 [shape=box, label="Projection: a@0, a@1, a@2"]
-		stage_4_0_0_0 [shape=box, label="CoalesceBatches [batchSize=4096]"]
-		stage_4_0_0_0_0 [shape=box, label="HashJoin
+		label = "Stage 4 [Resolved]";
+		stage_4_0 [shape=box, label="ShuffleWriter [0 partitions]"]
+		stage_4_0_0 [shape=box, label="MemoryExec"]
+		stage_4_0_0 -> stage_4_0
+	}
+	subgraph cluster4 {
+		label = "Stage 5 [Unresolved]";
+		stage_5_0 [shape=box, label="ShuffleWriter [48 partitions]"]
+		stage_5_0_0 [shape=box, label="Projection: a@0, a@1, a@2"]
+		stage_5_0_0_0 [shape=box, label="CoalesceBatches [batchSize=8192]"]
+		stage_5_0_0_0_0 [shape=box, label="HashJoin
 join_expr=a@1 = a@0
 filter_expr="]
-		stage_4_0_0_0_0_0 [shape=box, label="CoalesceBatches [batchSize=4096]"]
-		stage_4_0_0_0_0_0_0 [shape=box, label="HashJoin
-join_expr=a@0 = a@0
-filter_expr="]
-		stage_4_0_0_0_0_0_0_0 [shape=box, label="CoalesceBatches [batchSize=4096]"]
-		stage_4_0_0_0_0_0_0_0_0 [shape=box, label="UnresolvedShuffleExec [stage_id=1]"]
-		stage_4_0_0_0_0_0_0_0_0 -> stage_4_0_0_0_0_0_0_0
-		stage_4_0_0_0_0_0_0_0 -> stage_4_0_0_0_0_0_0
-		stage_4_0_0_0_0_0_0_1 [shape=box, label="CoalesceBatches [batchSize=4096]"]
-		stage_4_0_0_0_0_0_0_1_0 [shape=box, label="UnresolvedShuffleExec [stage_id=2]"]
-		stage_4_0_0_0_0_0_0_1_0 -> stage_4_0_0_0_0_0_0_1
-		stage_4_0_0_0_0_0_0_1 -> stage_4_0_0_0_0_0_0
-		stage_4_0_0_0_0_0_0 -> stage_4_0_0_0_0_0
-		stage_4_0_0_0_0_0 -> stage_4_0_0_0_0
-		stage_4_0_0_0_0_1 [shape=box, label="CoalesceBatches [batchSize=4096]"]
-		stage_4_0_0_0_0_1_0 [shape=box, label="UnresolvedShuffleExec [stage_id=3]"]
-		stage_4_0_0_0_0_1_0 -> stage_4_0_0_0_0_1
-		stage_4_0_0_0_0_1 -> stage_4_0_0_0_0
-		stage_4_0_0_0_0 -> stage_4_0_0_0
-		stage_4_0_0_0 -> stage_4_0_0
-		stage_4_0_0 -> stage_4_0
+		stage_5_0_0_0_0_0 [shape=box, label="CoalesceBatches [batchSize=8192]"]
+		stage_5_0_0_0_0_0_0 [shape=box, label="UnresolvedShuffleExec [stage_id=3]"]
+		stage_5_0_0_0_0_0_0 -> stage_5_0_0_0_0_0
+		stage_5_0_0_0_0_0 -> stage_5_0_0_0_0
+		stage_5_0_0_0_0_1 [shape=box, label="CoalesceBatches [batchSize=8192]"]
+		stage_5_0_0_0_0_1_0 [shape=box, label="UnresolvedShuffleExec [stage_id=4]"]
+		stage_5_0_0_0_0_1_0 -> stage_5_0_0_0_0_1
+		stage_5_0_0_0_0_1 -> stage_5_0_0_0_0
+		stage_5_0_0_0_0 -> stage_5_0_0_0
+		stage_5_0_0_0 -> stage_5_0_0
+		stage_5_0_0 -> stage_5_0
 	}
-	stage_1_0 -> stage_4_0_0_0_0_0_0_0_0
-	stage_2_0 -> stage_4_0_0_0_0_0_0_1_0
-	stage_3_0 -> stage_4_0_0_0_0_1_0
+	stage_1_0 -> stage_3_0_0_0_0_0
+	stage_2_0 -> stage_3_0_0_0_1_0
+	stage_3_0 -> stage_5_0_0_0_0_0_0
+	stage_4_0 -> stage_5_0_0_0_0_1_0
 }
-"#;
-        assert_eq!(expected, &dot);
+"#
+        .trim();
+        assert_eq!(dot.trim(), expected);
         Ok(())
     }
 
     #[tokio::test]
     async fn query_stage_optimized() -> Result<()> {
         let graph = test_graph_optimized().await?;
         let dot = ExecutionGraphDot::generate_for_query_stage(Arc::new(graph), 4)
```

Comment:
The plan has changed a lot due to optimizations DF ... I think looking at stage 3 or 5 here rather than 4 would be more interesting though (those have joins)

Reply:
I switched it to 3